text/0000-maybe-dangling.md

```diff
@@ -0,0 +1,268 @@
+  (On the pointee side, Rust used to have a `Unique` type, that still exists internally in the standard library, which was intended to provide `noalias` without any form of `dereferenceable`. It was deemed better to not expose this.)
+- Instead of saying that all fields of all compound types still must abide by the aliasing rules, we could restrict this to fields of `repr(transparent)` types.
+  That would solve the 2nd and 3rd example without any code changes.
+  It would make it impossible to package up multiple references (in a struct with multiple reference-typed fields) in a way that their aliasing guarantees are still in full force.
+  Right now, we actually *do* emit `noalias` for the 2nd and 3rd example, so codegen of existing types would have to be changed under this alternative.
+  It would not help for the first example.
+- Finally we could do nothing and declare all examples as intentional UB.
+  The 2nd and 3rd example could use `MaybeUninit` to pass around the closure / the buffer in a UB-free way.
+  That will however require `unsafe` code, and leaves `ManuallyDrop<Box<T>>` with its footgun (1st example).
+
+# Prior art
+[prior-art]: #prior-art
+
+The author cannot think of prior art in other languages; the issue arises because of Rust's unique combination of strong safety guarantees with low-level types such as `ManuallyDrop` that manage memory allocation in a very precise way.
+
+Inside Rust, we do have precedent for wrapper types altering language semantics; most prominently, there are `UnsafeCell` and `MaybeUninit`.
+Notice that `UnsafeCell` acts "behind references" while `MaybeDangling`, like `MaybeUninit`, acts "around references": `MaybeDangling<&T>` vs `&UnsafeCell<T>`.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- What should the type be called?
+  `MaybeDangling` is somewhat misleading since the *safety* invariant still requires everything to be dereferenceable, only the *validity* requirement of dereferenceability and noalias is relaxed.
+  This is a bit like `ManuallyDrop` which supports dropping via an `unsafe` function but its safety invariant says that the data is not dropped (so that it can implement `Deref` and `DerefMut` and a safe `into_inner`).
+  Furthermore, the type also allows maybe-aliasing references, not just maybe-dangling references.
+  Other possible names might be things like `InertPointers` or `SuspendedPointers`.
+- Should `MaybeDangling` implement `Deref` and `DerefMut` like `ManuallyDrop` does, or should accessing the inner data be more explicit since that is when the aliasing and dereferencability requirements do come back in full force?
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+- None that the author can think of -- this arguably closes a gap in our ability to express and manipulate the aliasing guarantees of types that are being passed around.
```

Comment:
Could you add a future possibility of allowing moving fields out of `MaybeDangling`, the same way `Box` currently works? It doesn't have to be the same mechanism, ideally we'd introduce a mechanism for any type to do this, instead of adding more magic.

(I know this is a very _future_ possibility, but still...)

Reply:
Since it looks like we'll land the semantics first and the exact mechanism to expose this can still be tweaked, I'll put that together with the attribute approach.

Reply:
Thanks <3